flow/flow_test.go

```diff
@@ -0,0 +1,198 @@
+	f.ID = r.ID
+	f.LoginRemember = r.Remember
+	f.LoginRememberFor = r.RememberFor
+	f.ACR = r.ACR
+	f.AMR = r.AMR
+	f.Subject = r.Subject
+	f.ForceSubjectIdentifier = r.ForceSubjectIdentifier
+	f.Context = r.Context
+	f.LoginWasUsed = r.WasHandled
+	f.LoginError = r.Error
+	f.RequestedAt = r.RequestedAt
+	f.LoginAuthenticatedAt = r.AuthenticatedAt
+}
+
+func (f *Flow) setConsentRequest(r consent.ConsentRequest) {
+	f.ConsentChallengeID = sqlxx.NullString(r.ID)
+	f.RequestedScope = r.RequestedScope
+	f.RequestedAudience = r.RequestedAudience
+	f.ConsentSkip = r.Skip
+	f.Subject = r.Subject
+	f.OpenIDConnectContext = r.OpenIDConnectContext
+	f.Client = r.Client
+	f.ClientID = r.ClientID
+	f.RequestURL = r.RequestURL
+	f.ID = r.LoginChallenge.String()
+	f.SessionID = r.LoginSessionID
+	f.ACR = r.ACR
+	f.AMR = r.AMR
+	f.Context = r.Context
+	f.ConsentWasHandled = r.WasHandled
+	f.ForceSubjectIdentifier = r.ForceSubjectIdentifier
+	f.ConsentVerifier = sqlxx.NullString(r.Verifier)
+	f.ConsentCSRF = sqlxx.NullString(r.CSRF)
+	f.LoginAuthenticatedAt = r.AuthenticatedAt
+	f.RequestedAt = r.RequestedAt
+}
+
+func (f *Flow) setHandledConsentRequest(r consent.HandledConsentRequest) {
+	f.ConsentChallengeID = sqlxx.NullString(r.ID)
+	f.GrantedScope = r.GrantedScope
+	f.GrantedAudience = r.GrantedAudience
+	f.ConsentRemember = r.Remember
+	f.ConsentRememberFor = &r.RememberFor
+	f.ConsentHandledAt = r.HandledAt
+	f.ConsentWasHandled = r.WasHandled
+	f.ConsentError = r.Error
+	f.RequestedAt = r.RequestedAt
+	f.LoginAuthenticatedAt = r.AuthenticatedAt
+}
+
+func TestFlow_GetLoginRequest(t *testing.T) {
+	t.Run("GetLoginRequest should set all fields on its return value", func(t *testing.T) {
+		f := Flow{}
+		expected := consent.LoginRequest{}
+		assert.NoError(t, faker.FakeData(&expected))
+		f.setLoginRequest(&expected)
+		actual := f.GetLoginRequest()
+		assert.Equal(t, expected, *actual)
+	})
+}
+
+func TestFlow_GetHandledLoginRequest(t *testing.T) {
+	t.Run("GetHandledLoginRequest should set all fields on its return value", func(t *testing.T) {
+		f := Flow{}
+		expected := consent.HandledLoginRequest{}
+		assert.NoError(t, faker.FakeData(&expected))
+		f.setHandledLoginRequest(&expected)
+		actual := f.GetHandledLoginRequest()
+		expected.LoginRequest = nil
+		actual.LoginRequest = nil
+		assert.Equal(t, expected, actual)
+	})
+}
+
+func TestFlow_NewFlow(t *testing.T) {
+	t.Run("NewFlow and GetLoginRequest should use all LoginRequest fields", func(t *testing.T) {
+		expected := &consent.LoginRequest{}
+		assert.NoError(t, faker.FakeData(expected))
+		actual := NewFlow(expected).GetLoginRequest()
+		assert.Equal(t, expected, actual)
+	})
+}
+
+func TestFlow_HandleLoginRequest(t *testing.T) {
+	t.Run(
+		"HandleLoginRequest should ignore RequestedAt in its argument and copy the other fields",
+		func(t *testing.T) {
+			f := Flow{}
+			assert.NoError(t, faker.FakeData(&f))
+			f.State = FlowStateLoginInitialized
```

Comment:
Consider adding & testing

```
f.SessionIDToken = ...
f.SessionAccessToken = ...
```

Reply:
I added it to `TestFlow_GetHandledConsentRequest`. Not sure about `TestFlow_HandleLoginRequest` - there should be no tokens yet?